### PR TITLE
AJ-636: Display Azure file URIs in WDS for Azure workspaces

### PR DESCRIPTION
**Pull request by aaronkanzer**
Relates to https://broadworkbench.atlassian.net/browse/AJ-636

Makes URIs into links in Azure Workspaces for Azure-specific & DRS URLs

Disables GS links in Azure Workspaces

Some follow-up work remains, which is captured in https://broadworkbench.atlassian.net/browse/AJ-852 & https://broadworkbench.atlassian.net/browse/AJ-871

### CODE REVIEW

File: src/components/UriViewer.js [hunk 1]
```diff
@@ -176,7 +177,7 @@ const UriViewer = _.flow(
 
   const loadMetadata = async () => {
     try {
-      if (isGs(uri)) {
+      if (isGs(uri) && false) {
```
Comment (cahrens) [inline src/components/UriViewer.js]: The false seems suspicious, but given that this is draft perhaps you are just testing.

File: src/components/data/data-utils.js [hunk 2]
```diff
@@ -115,8 +118,7 @@ export const renderDataCell = (attributeValue, workspace) => {
 
   const renderCell = datum => {
     const stringDatum = Utils.convertValue('string', datum)
-
-    return isUri(datum) ? h(UriViewerLink, { uri: datum, workspace }) : stringDatum
+    return isUri(datum, workspace.googleProject) ? h(UriViewerLink, { uri: datum, workspace }) : stringDatum
```
Comment (cahrens) [inline src/components/data/data-utils.js]: The preferred way to check if the workspace is a Google workspace is to use `isGoogleWorkspace` from `workspace-utils` (and similarly `isAzureWorkspace`).

File: src/components/UriViewer.js (hunk 1 above)
Comment (aaronkanzer) [reply src/components/UriViewer.js]: yeah this was a fragment of this still being in draft -- will remove the draft tag once this is ready for review!

File: src/components/data/data-utils.js (hunk 2 above)
Comment (aaronkanzer) [reply src/components/data/data-utils.js]: Updated this PR to use those functions!

**Comment by aaronkanzer** [comment]
resolving integration test....

**Comment by calypsomatic** [comment]
It's true it's not a valid DRS uri, but it is a valid azure storage container uri!   I copied this uri from a file I uploaded into the files area of the workspace so it seems like it should work.
image

File: integration-tests/utils/integration-helpers.js [hunk 3]
```diff
@@ -105,6 +105,7 @@ const deleteWorkspace = withSignedInPage(async ({ page, billingProject, workspac
 
 const withWorkspace = test => async options => {
   console.log('withWorkspace ...')
+  console.log(options)
```
Comment (davidangb) [inline integration-tests/utils/integration-helpers.js]: is this change necessary/appropriate?

File: src/components/UriViewer.js [hunk 4]
```diff
@@ -56,9 +56,11 @@ const isFilePreviewable = ({ size, ...metadata }) => {
   return !isBinary(metadata) && (isText(metadata) || (isImage(metadata) && size <= 1e9))
 }
 
-const isGs = uri => _.startsWith('gs://', uri)
+export const isGs = uri => _.startsWith('gs://', uri)
 
-const isDrs = uri => _.startsWith('dos://', uri) || _.startsWith('drs://', uri)
+export const isDrs = uri => _.startsWith('dos://', uri) || _.startsWith('drs://', uri)
+
+export const isAzureUri = uri => _.includes('blob.core.windows.net', uri)
```
Comment (davidangb) [inline src/components/UriViewer.js]: probably not a big deal, but this is a lot looser of a check than the `startsWith` checks for gs, dos, drs. This would match "ftp://somehost/somefile?downloadedFrom=blob.core.windows.net", for instance.

File: src/components/data/data-utils.test.js [hunk 5]
```diff
@@ -136,66 +152,91 @@ describe('renderDataCell', () => {
...
       const link = getByRole('link')
       expect(link).toHaveAttribute('href', 'drs://example.data.service.org/6cbffaae-fc48-4829-9419-1a2ef0ca98ce')
     })
+
+    it('renders links to Azure URLs', () => {
+      const { container, getByRole } = render(renderDataCell('https://sa226344b664da26ad6863.blob.core.windows.net/file1.cram', testAzureWorkspace))
+      expect(container).toHaveTextContent('https://sa226344b664da26ad6863.blob.core.windows.net/file1.cram')
+      const link = getByRole('link')
+      expect(link).toHaveAttribute('href', 'https://sa226344b664da26ad6863.blob.core.windows.net/file1.cram')
+    })
```
Comment (davidangb) [inline src/components/data/data-utils.test.js:239]: nice updates to these tests!

**Review by davidangb — CHANGES_REQUESTED** [review]
a couple comments inline - also, following up on Bria's comment/screenshot, all of the file-preview functionality will also need updates. Start with `isFilePreviewable` and `PreviewContent` in `UriViewer.js` to see that code.

File: integration-tests/utils/integration-helpers.js (hunk 3 above)
Comment (aaronkanzer) [reply integration-tests/utils/integration-helpers.js]: Removed! Was an artifact of just inspecting the test!

File: src/components/UriViewer.js (hunk 4 above)
Comment (aaronkanzer) [reply src/components/UriViewer.js]: Agreed, this is a tough one to predict/be more specific as the URIs start with some unique identifier, rather than a URL prefix -- e.g. `sa226344b664da26ad6863` is the prefix to a sample Azure URI

Comment (nawatts) [reply src/components/UriViewer.js]: Is there a regex we could use for this? Something along the lines of `https://(.+).blob.core.windows.net`?

File: src/components/UriViewer.js [hunk 6]
```diff
@@ -183,6 +187,25 @@ const UriViewer = _.flow(
         })
         const metadata = await loadObject(googleProject, bucket, name)
         setMetadata(metadata)
+      } else if (isAzureUri(uri)) {
+        const provider = AzureBlobStorageFileBrowserProvider(
+          { workspaceId: workspace.workspace.workspaceId, pageSize: 1 }
+        )
+        const getFile = await provider.getFilesInDirectory('').then(files => {
```
Comment (nawatts) [inline src/components/UriViewer.js]: I'm confused here... It looks like this is always getting metadata for the first file at the top level of the storage container, not the file referenced by `uri`?

File: src/components/UriViewer.js [hunk 7]
```diff
@@ -183,6 +187,25 @@ const UriViewer = _.flow(
...
+          return files.items[0]
+        })
+        await provider.getDownloadUrlForFile(getFile.path).then(url => {
```
Comment (nawatts) [inline src/components/UriViewer.js]: Style nit: Instead of handling the result in a `then` callback, this could be written as
```js
const url = await provider.getDownloadUrlForFile(...)
const metadata = { ... }
...
```

File: src/components/UriViewer.js [hunk 8]
```diff
@@ -277,7 +301,7 @@ const UriViewer = _.flow(
         div({ style: { fontSize: 10 } }, ['* Estimated. Download cost may be higher in China or Australia.'])
       ])],
       () => h(Fragment, [
-        isGs(uri) ? 'Loading metadata...' : 'Resolving DOS object...',
+        isGs(uri) ? 'Loading metadata...' : isAzureUri(uri) ? 'Retrieving Azure file...' : isDrs(uri) ? 'Resolving DRS object...' : 'Retrieving file...',
```
Comment (nawatts) [inline src/components/UriViewer.js]: Instead of nesting ternaries, prefer `Utils.cond`.

File: src/components/data/data-utils.js [hunk 9]
```diff
@@ -61,13 +62,19 @@ export const getDownloadCommand = (fileName, gsUri, accessUrl) => {
...
 }
 
 export const getUserProjectForWorkspace = async workspace => (workspace && await canUseWorkspaceProject(workspace)) ?
-  workspace.workspace.googleProject :
+  isGoogleWorkspace(workspace) :
   requesterPaysProjectStore.get()
 
-const isUri = datum => _.startsWith('gs://', datum) || _.startsWith('dos://', datum) || _.startsWith('drs://', datum)
+const isUri = (datum, workspace) => (isGoogleWorkspace(workspace) &&
+  isGs(datum)) || isDrs(datum) || (isAzureWorkspace(workspace) && (isAzureUri(datum)))
```
Comment (nawatts) [inline src/components/data/data-utils.js]: I think this change makes the function name misleading. A string could be a URI regardless of which type of workspace it's in. Is the new meaning of this something like `isViewableUri`?

Comment (nawatts) [reply src/components/data/data-utils.js]: Yeah, it looks like the usage of this is:
```js
return isUri(datum, workspace) ? h(UriViewerLink, { uri: datum, workspace }) : stringDatum
```

I'd vote for changing the name of this function. Even without this change, it's confusing since a reader could reasonably assume `isUri('https://example.com/file.txt')` would return true. Not necessary to add in this PR though.

File: src/libs/ajax/file-browser-providers/AzureBlobStorageFileBrowserProvider.ts [hunk 10]
```diff
@@ -11,7 +11,7 @@ export interface AzureBlobStorageFileBrowserProviderParams {
 }
 
 type AzureBlobStorageFileBrowserProviderGetPageParams<T> = {
-  isFirstPage: boolean
+  isFirstPage?: boolean
```
Comment (nawatts) [inline src/libs/ajax/file-browser-providers/AzureBlobStorageFileBrowserProvider.ts]: Is there somewhere in here that needs to be changed to set a default value for `isFirstPage` to accommodate this change?

File: src/libs/ajax/file-browser-providers/AzureBlobStorageFileBrowserProvider.ts [hunk 11]
```diff
@@ -29,7 +29,6 @@ interface BlobListRequestOptions {
 
 const AzureBlobStorageFileBrowserProvider = ({ workspaceId, pageSize = 1000 }: AzureBlobStorageFileBrowserProviderParams): FileBrowserProvider => {
   const storageDetailsPromise = AzureStorage().details(workspaceId)
-
```
Comment (nawatts) [inline src/libs/ajax/file-browser-providers/AzureBlobStorageFileBrowserProvider.ts]: Why delete the blank lines? I find them useful for grouping sections of code.

File: src/components/UriViewer.js (hunk 7 above)
Comment (nawatts) [reply src/components/UriViewer.js]: More importantly, this line is where using a file browser provider is problematic. `getDownloadUrlForFile` takes a _path_, not a _URL_. This is because file browser providers can assume that any files they will have to deal with are in the workspace bucket / storage container. That assumption isn't true for data tables.

For files in another storage container, this would return an incorrect URL (because file browser providers construct the download URL using the workspace bucket/storage container). Even worse, if file is in another storage container and the workspace storage container has a file at the same path, the download link would appear to work but it would download a different file than expected.

**Review by nawatts — CHANGES_REQUESTED** [review]
I don't think file browser providers are particularly useful here. File browser provider methods take _paths_, not _URLs_. This is because file browser providers can assume that any files they will have to deal with are in the workspace bucket / storage container. That assumption isn't true for data tables.

I suppose a file browser provider could be used after checking if the file is in the workspace bucket / storage container. But since data tables would need a general solution for files in any bucket / storage container anyway, I'm not sure if it's worth using a file browser provider for that specific case.

File: src/components/UriViewer.js [hunk 12]
```diff
@@ -56,9 +57,14 @@ const isFilePreviewable = ({ size, ...metadata }) => {
...
 }
 
-const isGs = uri => _.startsWith('gs://', uri)
+export const isGs = uri => _.startsWith('gs://', uri)
 
-const isDrs = uri => _.startsWith('dos://', uri) || _.startsWith('drs://', uri)
+export const isDrs = uri => _.startsWith('dos://', uri) || _.startsWith('drs://', uri)
+
+export const isAzureUri = uri => {
+  const azureRegex = RegExp('https://(.+).blob.core.windows.net')
```
Comment (nawatts) [inline src/components/UriViewer.js]: This could probably be tightened up a bit. For example, checking that this pattern appears at the start of the `uri` string. Also, I'm guessing there's a limited set of characters allowed for the storage container ID.

File: src/components/UriViewer.js [hunk 13]
```diff
@@ -277,7 +284,7 @@ const UriViewer = _.flow(
         div({ style: { fontSize: 10 } }, ['* Estimated. Download cost may be higher in China or Australia.'])
       ])],
       () => h(Fragment, [
-        isGs(uri) ? 'Loading metadata...' : 'Resolving DOS object...',
+        isGs(uri) ? 'Loading metadata...' : 'Resolving DRS object...',
```
Comment (nawatts) [inline src/components/UriViewer.js]: Does this need to handle Azure URIs?

File: src/components/data/data-utils.js [hunk 14]
```diff
@@ -64,10 +65,11 @@ export const getDownloadCommand = (fileName, gsUri, accessUrl) => {
 }
 
 export const getUserProjectForWorkspace = async workspace => (workspace && await canUseWorkspaceProject(workspace)) ?
-  workspace.workspace.googleProject :
+  isGoogleWorkspace(workspace) :
```
Comment (nawatts) [inline src/components/data/data-utils.js]: This looks wrong (maybe an artifact of converting checks for Google workspaces to `isGoogleWorkspace`). `getUserProjectForWorkspace` should return a string (the Google project), not a boolean.

**Comment by aaronkanzer** [comment]
@nawatts Just as a heads-up: @bethsheets and I limited the scope here and made a follow-up ticket: https://broadworkbench.atlassian.net/browse/AJ-871

Just wanted to let you know since I've simplified a lot of changes here -- will link your review to the next JIRA ticket 🙌

File: src/components/UriViewer.js (hunk 4 above)
Comment (aaronkanzer) [reply src/components/UriViewer.js]: Updated!

**Comment by aaronkanzer** [comment]
> I don't think file browser providers are particularly useful here. File browser provider methods take _paths_, not _URLs_. This is because file browser providers can assume that any files they will have to deal with are in the workspace bucket / storage container. That assumption isn't true for data tables.
> 
> I suppose a file browser provider could be used after checking if the file is in the workspace bucket / storage container. But since data tables would need a general solution for files in any bucket / storage container anyway, I'm not sure if it's worth using a file browser provider for that specific case.

Spoke in-person w/ @nawatts -- decided to not use the file provider here for now

**Comment by nawatts** [comment]
@aaronkanzer 👍 Does that mean this PR is ready for re-review?

File: src/components/UriViewer.js (hunk 13 above)
Comment (aaronkanzer) [reply src/components/UriViewer.js]: Removed as part of https://github.com/DataBiosphere/terra-ui/pull/3780#issuecomment-1433494314

File: src/components/UriViewer.js (hunk 12 above)
Comment (aaronkanzer) [reply src/components/UriViewer.js]: I updated it to ensure that the regex starts at the beginning of the uri provided

It might be tricky to enforce any further: https://learn.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#blob-names

File: src/components/data/data-utils.js (hunk 14 above)
Comment (aaronkanzer) [reply src/components/data/data-utils.js]: Ah good catch -- meant to remove this

**Comment by aaronkanzer** [comment]
> @aaronkanzer 👍 Does that mean this PR is ready for re-review?

yes! thanks again

File: src/components/data/data-utils.js [hunk 15]
```diff
@@ -67,7 +68,8 @@ export const getUserProjectForWorkspace = async workspace => (workspace && await
   workspace.workspace.googleProject :
   requesterPaysProjectStore.get()
 
-const isUri = datum => _.startsWith('gs://', datum) || _.startsWith('dos://', datum) || _.startsWith('drs://', datum)
+const isViewableUri = (datum, workspace) => (isGoogleWorkspace(workspace) &&
+  isGs(datum)) || isDrs(datum) || (isAzureWorkspace(workspace) && (isAzureUri(datum)))
```
Comment (nawatts) [inline src/components/data/data-utils.js]: Style nit: This expression is kind of hard to read. Maybe it could be tidied up a bit (IDK if ESLint will accept this exact suggestion)?
```suggestion
const isViewableUri = (datum, workspace) =>
  (isGoogleWorkspace(workspace) && isGs(datum)) ||
  (isAzureWorkspace(workspace) && isAzureUri(datum)) ||
  isDrs(datum)
```

**Review by nawatts — APPROVED** [review]

File: src/components/data/data-utils.js (hunk 15 above)
Comment (aaronkanzer) [reply src/components/data/data-utils.js]: Updated here!

File: src/components/data/data-utils.js (hunk 9 above)
Comment (aaronkanzer) [reply src/components/data/data-utils.js]: Updated naming!

File: src/components/UriViewer.js [hunk 16]
```diff
@@ -292,9 +297,9 @@ export const UriViewerLink = ({ uri, workspace }) => {
       href: uri,
       onClick: e => {
         e.preventDefault()
-        setModalOpen(true)
+        setModalOpen(!isAzureUri(uri))
```
Comment (nawatts) [inline src/components/UriViewer.js]: ? For Azure URLs, why do we render this link if it can't be used? This would appear broken if clicking on the link does nothing.

Comment (aaronkanzer) [reply src/components/UriViewer.js]: After chatting with @bethsheets the other day, I think the goal was this to be indicative of being type `File` for Cromwell workflows for the time being until https://broadworkbench.atlassian.net/browse/AJ-871 is implemented

Comment (nawatts) [reply src/components/UriViewer.js]: I still think clicking the link should do something. This would appear broken to a user. Until AJ-871, can we open the modal but only display limited information in the modal, even if it's just the file name and "Download button coming soon"?

Comment (bethsheets) [reply src/components/UriViewer.js]: I think clicking the link and having a modal with limited information sounds like a better user experience. I suggest the modal say "Download functionality is coming soon."

Comment (aaronkanzer) [reply src/components/UriViewer.js]: Thanks both -- will update here accordingly then!

**Review by jordansafer — APPROVED** [review]
Thank you for all your work on this!

File: src/components/UriViewer.js (hunk 16 above)
Comment (aaronkanzer) [reply src/components/UriViewer.js]: Screenshot 2023-02-17 at 11 51 00 AM

I've updated the code to provide a modal that looks as such -- let me know if this works!

**Comment by bethsheets** [comment]
LGTM!

File: src/components/UriViewer.js [hunk 17]
```diff
@@ -56,9 +56,14 @@ const isFilePreviewable = ({ size, ...metadata }) => {
...
-const isGs = uri => _.startsWith('gs://', uri)
+export const isGs = uri => _.startsWith('gs://', uri)
 
-const isDrs = uri => _.startsWith('dos://', uri) || _.startsWith('drs://', uri)
+export const isDrs = uri => _.startsWith('dos://', uri) || _.startsWith('drs://', uri)
+
+export const isAzureUri = uri => {
+  const azureRegex = RegExp('^https://(.+).blob.core.windows.net')
+  return azureRegex.test(uri)
+}
```
Comment (davidangb) [inline src/components/UriViewer.js]: nit: move this `const azureRegex` out of `isAzureUri()` and up to the root level of `UriViewer.js` so we're not re-creating a new `RegExp` for every invocation of `isAzureUri()`

```suggestion
const azureRegex = RegExp('^https://(.+).blob.core.windows.net')

export const isAzureUri = uri => {
  return uri.match(azureRegex)
}
```

note that I've also changed from `RegExp.test()` to `String.match()` to avoid problems with regex reuse: https://blog.rebased.pl/2021/03/16/a-regular-expression-surprise-in-javascript.html (we don't use the /g flag so it wouldn't hit us, but still)

File: src/components/UriViewer.js [hunk 18]
```diff
@@ -176,27 +181,32 @@ const UriViewer = _.flow(
 
   const loadMetadata = async () => {
     try {
+      let metadata = {}
```
Comment (davidangb) [inline src/components/UriViewer.js]: could you do this without a `let`, keeping `metadata` as a const? There are three cases for the if conditions below - you could create a `const metadata` inside each condition and call `setMetadata` inside each condition. Note that right now the first case calls `setMetadata` internally but the second and third don't - the second and third rely on the `setMetadata` that happens outside the if, which is redundant for the first case.

Or, you could have `const metadata = ` the return value of the if statement, then call `setMetadata` outside the if.

File: src/components/UriViewer.js [hunk 19]
```diff
@@ -216,7 +226,7 @@ const UriViewer = _.flow(
     okButton: 'Done'
   }, [
     Utils.cond(
-      [loadingError, () => h(Fragment, [
+      [!isAzureUri(uri) && loadingError, () => h(Fragment, [
```
Comment (davidangb) [inline src/components/UriViewer.js]: there are so many `!isAzureUri(…)` conditionals peppered throughout here. How about something that doesn't require so many changes? For instance:
```
if (isAzureUri(uri)) {
    return h(Modal, {
      onDismiss,
      title: 'File Details',
      showCancel: false,
      showX: true,
      okButton: 'Done'
    }, ['Download functionality for Azure files coming soon'])
  } else {
		… return original Modal …
}
```

**Review by davidangb — COMMENTED** [review]
refactoring suggestions inline, thanks for pushing on this!

File: src/components/UriViewer.js (hunk 16 above)
Comment (davidangb) [reply src/components/UriViewer.js]: note that the logic for what gets a link here is not exactly the same as what WDS uses for its `File` datatype - specifically support for dos:// uris. If we truly wanted these to be in sync we should probably ask WDS for its datatype and display a link based on the column's WDS definition, instead of inspecting the content directly. That can/should be for a future PR.

File: src/components/UriViewer.js (hunk 17 above)
Comment (aaronkanzer) [reply src/components/UriViewer.js]: great point here -- updated!

File: src/components/UriViewer.js (hunk 18 above)
Comment (aaronkanzer) [reply src/components/UriViewer.js]: Realized we no longer need to make changes in this function per your suggestion in https://github.com/DataBiosphere/terra-ui/pull/3780#discussion_r1110306666

File: src/components/UriViewer.js (hunk 19 above)
Comment (aaronkanzer) [reply src/components/UriViewer.js]: Great call here -- refactored!

**Review by davidangb — APPROVED** [review]
thank you!
